base/collect_database.go

```diff
@@ -0,0 +1,311 @@
+
+	return value, true
+}
+
+// run executes this task
+func (t *collectDatabaseTask) Run(exp *Experiment) error {
+	// validate inputs
+	var err error
+
+	err = t.validateInputs()
+	if err != nil {
+		return err
+	}
+
+	// initialize defaults
+	t.initializeDefaults()
+
+	// get current directory path
+	path, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	// collect all files paths in current directory that ends with metrics.yaml
+	metricFilePaths := []string{}
+	err = filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !info.IsDir() && strings.HasSuffix(path, "metrics.yaml") {
+			metricFilePaths = append(metricFilePaths, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// collect metrics for all metric files and versionInfos
+	for _, metricFilePath := range metricFilePaths {
+		for _, versionInfo := range t.With.VersionInfo {
+			// add ElapsedTime
+			elapsedTime, err := getElapsedTime(versionInfo, exp)
+			if err != nil {
+				return err
+			}
+			versionInfo["ElapsedTime"] = elapsedTime
+
+			fmt.Println(elapsedTime)
+
+			// finalize metrics template
+			template, err := template.ParseFiles(metricFilePath)
+			if err != nil {
+				return err
+			}
+			var buf bytes.Buffer
+			err = template.Execute(&buf, versionInfo)
+			if err != nil {
+				return err
+			}
+			var metrics CollectDatabaseTemplate
+			err = yaml.Unmarshal(buf.Bytes(), &metrics)
+			if err != nil {
+				return err
+			}
+
+			for _, metric := range metrics.Metrics {
+				// perform database query and extract metric value
+				value, ok := queryDatabaseAndGetValue(metrics, metric)
+
+				// check if there were any issues querying database and extracting value
+				if !ok {
+					continue
+				}
+
+				// do not save value if it has no value
+				if value == nil {
+					log.Logger.Error("could not extract non-nil value for metric ", metric.Name, ": ", err)
+					continue
+				}
+
+				// determine metric name
+				pathTokens := strings.Split(metricFilePath, "/")
+				fileNameWithExtension := pathTokens[len(pathTokens)-1]
+				fileNameTokens := strings.Split(fileNameWithExtension, ".metrics.yaml")
+				fileName := fileNameTokens[0]
+				metricName := fileName + "/" + metric.Name
+
+				// determine metric type
+				var metricType MetricType
+				if metric.Type == "gauge" {
+					metricType = GaugeMetricType
+				} else if metric.Type == "counter" {
+					metricType = CounterMetricType
+				}
+
+				// finalize metric data
+				mm := MetricMeta{
+					Description: metric.Description,
+					Type:        metricType,
+					Units:       &metric.Units,
+				}
+
+				// convert value to float
+				valueString := fmt.Sprint(value)
+				floatValue, err := strconv.ParseFloat(valueString, 64)
+				if err != nil {
+					log.Logger.Error("could not parse string \""+valueString+"\" to float:", err)
+					continue
+				}
+
+				exp.Result.Insights.updateMetric(metricName, mm, 0, floatValue)
```

Comment:
Shouldn't this be `err = `exp.Result....?

Reply:
Good catch.